fl4health/losses/contrastive_loss.py

```diff
@@ -86,3 +87,75 @@ def forward(
         labels = torch.zeros(features.size(0)).to(self.device).long()
 
         return self.cross_entropy_function(logits, labels)
+
+
+class NtXentLoss(nn.Module):
+    def __init__(self, device: torch.device, temperature: float = 0.05) -> None:
+        """
+        Implementation of Normalized Temperature-Scaled Cross Entropy Loss (NT-Xent) proposed in
+        https://papers.nips.cc/paper_files/paper/2016/hash/6b180037abbebea991d8b1232f8a8ca9-Abstract.html
+        and notably used in SimCLR (https://arxiv.org/pdf/2002.05709) and FedSimCLR as proposed in Fed-X
+        (https://arxiv.org/pdf/2207.09158).
+
+        NT-Xent is a contrastive loss in which each feature has a positive pair and the rest of the features
+        are considered negative. It is computed based on the similarity of positive pairs relatice to negative
+        pairs.
+
+        Args:
+            device (torch.device): device to use for computation
+            temperature (float): temperature to scale the logits
+        """
+        super().__init__()
+        self.device = device
+        self.temperature = temperature
+
+    def forward(self, features: torch.Tensor, transformed_features: torch.Tensor) -> torch.Tensor:
+        """
+        Compute the contrastive loss based on the features and transformed_features. Given N features
+        and N transformed_features per batch, features[i] and transformed_features[i] are positive pairs
+        and the remaining 2N - 2 are negative pairs.
```

Comment:
Not at all! So we have N inputs and then N transformed versions of the inputs that we consider targets. So in each batch that is 2N samples. For each sample, one of the other samples is positive. This leaves the remaining 2N - 2 samples as negatives. 2N - 2 because we are subtracting out the sample and its positive pair. Let me know if I can clarify the documentation at all!

Reply:
Ah right. I was thinking of the whole similarity matrix minus the diagonal, but that's not quite right. This wording is fine 🙂